File: src/learning-items/study-plan.ts
import type { LearningItem } from './types'

export type StudyPreferences = { daily_target: number; max_new_items: number }
export type PlannedItem = {
  was_new: boolean
  is_extra: boolean
  item: LearningItem
}
export type DailyStudyPlan = StudyPreferences & {
  local_date: string
  space_id: number | null
  completed_count: number
  total_count: number
  extra_completed_count: number
  can_study_more: boolean
  items: PlannedItem[]
}

export function studyPreferencesError(target: string, newItems: string): string {
  const daily = Number(target)
  const maximumNew = Number(newItems)
  if (!target.trim() || !Number.isInteger(daily) || daily < 1 || daily > 10000) {
    return 'Enter a daily study target between 1 and 10,000.'
  }
  if (!newItems.trim() || !Number.isInteger(maximumNew) || maximumNew < 0 || maximumNew > daily) {
    return 'Maximum new items must be between 0 and your daily study target.'
  }
  return ''
}

File: src-tauri/src/models/study_plan.rs
use super::learning_item::LearningItem;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(deny_unknown_fields)]
pub struct StudyPreferences {
    pub daily_target: i64,
    pub max_new_items: i64,
}

#[derive(Debug, Serialize)]
pub struct PlannedItem {
    pub was_new: bool,
    pub is_extra: bool,
    pub item: LearningItem,
}

#[derive(Debug, Serialize)]
pub struct DailyStudyPlan {
    pub local_date: String,
    pub daily_target: i64,
    pub max_new_items: i64,
    pub space_id: Option<i64>,
    pub completed_count: i64,
    pub total_count: i64,
    pub extra_completed_count: i64,
    pub can_study_more: bool,
    pub items: Vec<PlannedItem>,
}
